### PR TITLE
Fix declarative defaults for aliases

### PR DESCRIPTION
**Pull request by nikhita**
Fixes https://github.com/kubernetes/gengo/issues/235
For https://github.com/kubernetes/kubernetes/issues/104207

This commit ensures that the alias types are not unpacked to the underlying type in defaulter-gen. Instead the alias type name is carried into the generated code.

For the example described in #235:

Adding the declarative default as:

```go
// +default=ref(k8s.io/api/core/v1.ServiceInternalTrafficPolicyCluster)
InternalTrafficPolicy *ServiceInternalTrafficPolicy `json:"internalTrafficPolicy,omitempty" protobuf:"bytes,22,opt,name=internalTrafficPolicy"`
```

would generate code like:

```go
if in.Spec.InternalTrafficPolicy == nil {
  var ptrVar1 v1.ServiceInternalTrafficPolicy = v1.ServiceInternalTrafficPolicyCluster
  in.Spec.InternalTrafficPolicy = &ptrVar1
}
```

/cc @apelisse @thockin 
/assign @apelisse

### CODE REVIEW

File: examples/defaulter-gen/generators/defaulter.go [hunk 1]
```diff
@@ -562,8 +558,20 @@ func populateDefaultValue(node *callNode, t *types.Type, tags string, commentLin
 		node.markerOnly = true
 	}
 
-	node.defaultIsPrimitive = t.IsPrimitive()
```
Comment (MadhavJivrajani) [inline examples/defaulter-gen/generators/defaulter.go:565]: Unrelated diff? Or is `node` or `t` being effected by the added codepath?

**Review by MadhavJivrajani — COMMENTED** [review]
This looks great, thank you!

File: examples/defaulter-gen/generators/defaulter.go (hunk 1 above)
Comment (nikhita) [reply examples/defaulter-gen/generators/defaulter.go:565]: This line was moved to line 574. The diff just renders as this line being removed :/

**Comment by nikhita** [comment]
/hold
tests aren't running in CI, created https://github.com/kubernetes/gengo/pull/240 to fix it

**Comment by apelisse** [comment]
cc @alexzielenski Want to take a look?

File: examples/defaulter-gen/generators/defaulter.go [hunk 2]
```diff
@@ -466,9 +464,7 @@ func getNestedDefault(t *types.Type) string {
 		if len(defaultMap) == 1 && defaultMap[0] != "" {
 			return defaultMap[0]
 		}
-		if t.Kind == types.Alias {
-			t = t.Underlying
-		} else if t.Kind == types.Pointer {
+		if t.Kind == types.Pointer {
```
Comment (alexzielenski) [inline examples/defaulter-gen/generators/defaulter.go:467]: Changing this logic seems to remove the ability for a default attached to an alias's underlying to be used. Is that correct?

**Review by alexzielenski — CHANGES_REQUESTED** [review]
had a few questions about LocalNameOf, as well as some other small details

File: examples/defaulter-gen/generators/defaulter.go [hunk 3]
```diff
@@ -562,8 +558,20 @@ func populateDefaultValue(node *callNode, t *types.Type, tags string, commentLin
...
+	if t.Kind == types.Alias {
+		parsedName := types.ParseFullyQualifiedName(t.Name.String())
+		tracker := namer.NewDefaultImportTracker(types.Name{})
+		localPackage := generator.GolangTrackerLocalName(&tracker, t.Name)
+		if len(localPackage) > 0 {
+			node.defaultType = localPackage + "." + parsedName.Name
+		} else {
+			node.defaultType = parsedName.Name
+		}
+		node.underlying = t.Underlying
```
Comment (alexzielenski) [inline examples/defaulter-gen/generators/defaulter.go:571]: Does this work when there is an alias of an alias? I think its worth adding a test for that case and a depth of 3 for good measure.

File: examples/defaulter-gen/generators/defaulter.go (hunk 2 above)
Comment (thockin) [reply examples/defaulter-gen/generators/defaulter.go:467]: I want to argue that this FEELS correct (the same way that `type Y X` - Y doesn't inherit X methods, it shouldn't inherit defaults), but I don't know if that would break anything in k/k without actually trying.

**Comment by thockin** [comment]
cat_begging

**Comment by apelisse** [comment]
/close
Replaced by #246
